src/exact.rs

```diff
@@ -0,0 +1,144 @@
+use futures::stream::Stream;
+use pin_project::pin_project;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+/// An analogue of [`ExactSizeIterator`] for [`Stream`].
+/// This behaves exactly as you might expect based on the documentation of
+/// [`ExactSizeIterator`].
+///
+/// [`ExactSizeIterator`]: std::iter::ExactSizeIterator
+/// [`Stream`]: futures::stream::Stream
+pub trait ExactSizeStream: Stream {
+    /// Return the length of the stream that remains.
+    fn len(&self) -> usize {
+        let (lower, upper) = self.size_hint();
+        assert_eq!(upper, Some(lower));
+        lower
+    }
+
+    /// Return whether there are values remaining in the stream.
+    fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+#[pin_project]
+pub struct FixedLength<S> {
+    #[pin]
+    inner: S,
+    len: usize,
+}
+
+impl<S: Stream> FixedLength<S> {
+    /// Create a new fixed-length stream.
+    ///
+    /// Note that this is safe on the same basis that `ExactSizeIterator` is safe in
+    /// that code cannot rely on the claims of this type for ensuring safety of code.
+    #[allow(dead_code)] // TODO - use in infra
+    pub fn new(inner: S, len: usize) -> Self {
+        Self { inner, len }
+    }
+}
+
+impl<S: Stream> Stream for FixedLength<S> {
+    type Item = S::Item;
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        let res = this.inner.as_mut().poll_next(cx);
+        if let Poll::Ready(v) = &res {
+            if v.is_some() {
+                *this.len = (*this.len).wrapping_sub(1);
```

Comment:
I would also add a debug_assert here, it feels the same as providing len that is greater than actual stream length

Reply:
My theory was that any error will be caught at the end, so saving the assertions is fine.  Are you concerned that short streams (where we think the length is longer than it really is) need to be signaled sooner?  (A check here will never catch long streams.)

Reply:
yea I was thinking in our setting something is off if streams are short. At least in debug builds/tests we should always provide the exact length, I can't think of a valid use case where the mismatch is expected - although I could be missing something